src/main.rs

```diff
@@ -0,0 +1,291 @@
+#![no_std]
+#![no_main]
+#![feature(type_alias_impl_trait)]
+#[allow(arithmetic_overflow)]
+
+// use defmt::*;
+use embassy_executor::Spawner;
+use embassy_stm32::adc::Adc;
+use embassy_stm32::gpio::{Level, Input, Output, Pull, Speed};
+use embassy_stm32::usart::{Config, UartTx};
+use embassy_time::{Timer,Instant,Delay};
+use embassy_stm32::gpio::Level::{High,Low};
+
+use heapless::String;
+use {defmt_rtt as _, panic_probe as _};
+use ringbuffer::{ConstGenericRingBuffer, RingBuffer};
+
+use embassy_sync::signal::Signal;
+use embassy_sync::blocking_mutex::raw::CriticalSectionRawMutex;
+use static_cell::StaticCell;
+
+mod c5412;
+
+pub static C5412PINS_INST: StaticCell<c5412::C5412Pins> = StaticCell::new();
+
+const CRAZY_HI: i32 =3000;
+const CRAZY_LO: i32 =-1000;
+const DC_ALPHA: f64 = 1.0/1000.0;
+const LP_ALPHA: f64 = 1.0/100.0;
+const THRESHOLD_ALPHA_UP: f64 = 1.0/100.0;
+const THRESHOLD_ALPHA_DN: f64 = 1.0/2000.0;
+const PEAK_DELAY: usize = 200;
+
+const ABOVE_SIZE: usize = 200;
+const BELOW_SIZE: usize = 200;
+
+const DUMP_MODE: bool = true;
+
+struct Hr {
```

Comment:
I would pull this whole thing out into another file.
Including the task loop that runs it.
Leave the global statics here, and just pass the references

Could consider leaving the task here isntead

Reply:
did both, pulling everything but the task into another file